src/python_testing/TC_DGWIFI_2_1.py

```diff
@@ -0,0 +1,375 @@
+        asserts.assert_true(value is None or self.is_valid_uint_value(value, max_size=16),
+                            f"{field_name} should be a uint16 or NULL.")
+
+    def assert_valid_uint8(self, value, field_name):
+        """Asserts that the value is a valid uint16 or None (if attribute can return NULL)."""
+        asserts.assert_true(value is None or self.is_valid_uint_value(value, max_size=8),
+                            f"{field_name} should be a uint8 or NULL.")
+
+    async def read_dgwifi_attribute_expect_success(self, endpoint, attribute):
+        cluster = Clusters.Objects.WiFiNetworkDiagnostics
+        return await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=attribute)
+
+    def desc_TC_DGWIFI_2_1(self) -> str:
+        """Returns a description of this test."""
+        return "[TC-DGWIFI-2.1] Wi-Fi Network Diagnostics Attributes with Server as DUT"
+
+    def pics_TC_DGWIFI_2_1(self) -> list[str]:
+        return ["DGWIFI.S"]
+
+    def steps_TC_DGWIFI_2_1(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commission DUT to TH (already done)", is_commissioning=True),
+            TestStep(2, "Read BSSID attribute"),
+            TestStep(3, "Read SecurityType attribute"),
+            TestStep(4, "Read WiFiVersion attribute"),
+            TestStep(5, "Read ChannelNumber attribute"),
+            TestStep(6, "Read RSSI attribute"),
+            TestStep(7, "Read BeaconLostCount attribute"),
+            TestStep(8, "Read BeaconRxCount attribute"),
+            TestStep(9, "Read PacketMulticastRxCount attribute"),
+            TestStep(10, "Read PacketMulticastTxCount attribute"),
+            TestStep(11, "Read PacketUnicastRxCount attribute"),
+            TestStep(12, "Read PacketUnicastTxCount attribute"),
+            TestStep(13, "Read CurrentMaxRate attribute"),
+            TestStep(14, "Read OverrunCount attribute"),
+        ]
+        return steps
+
+    # ---------------------------
+    # Main Test Routine
+    # ---------------------------
+    @async_test_body
+    async def test_TC_DGWIFI_2_1(self):
+        endpoint = self.get_endpoint(default=0)
+
+        #
+        # STEP 1: Commissioning (already done)
+        #
+        self.step(1)
+        # Typically, we assume commissioning was performed by harness scripts.
+        attributes = Clusters.WiFiNetworkDiagnostics.Attributes
+
+        #
+        # STEP 2: TH reads Bssid attribute
+        #
+        self.step(2)
+        bssid = await self.read_dgwifi_attribute_expect_success(endpoint=endpoint, attribute=attributes.Bssid)
+        # If the interface is not configured or not operational, a None might be returned
+        if bssid is not None:
+            asserts.assert_true(isinstance(bssid, Nullable),
+                                "BSSID must be of type 'Nullable' when not None.")
+
+            if bssid is not NullValue:
+                bssid_value = bssid.Value
```

Comment:
why the separate assign? Could we make `assert_valid_bssid` also accept NullValue ? It already seems to accept none, but we check the none in advance.

How about having all the checks in the _assert_valid_bssid logic instead?